Replace docker-compose.yml file reading with config command

genezys opened the pull request:
- Replace the file reading with the command `docker-compose config` that will use all configuration files (extends, overrides) & will resolve relative volume paths
- Tweak regexp to only parse volumes starting with a / since paths are resolved by docker-compose, this prevents named volumes from being parsed

Fix #64 and #168

genezys committed:
Replace docker-compose.yml file reading with config command

- Replace the file reading with the command `docker-compose config` that will use all configuration files (extends, overrides) & will resolve relative volume paths
- Tweak regexp to only parse volumes starting with a / since paths are resolved by docker-compose, this prevents named volumes from being parsed

genezys commented:
All tests pass on my machine. 

I did not add tests for named volumes, it may be a good idea to add some but I do not really know where to start.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -992,28 +992,32 @@ function instructions {
-        else
-          in_volumes_block=false
+  local yaml_file_param=""
+  if [ ! -z "${yaml_file_path}" ]; then
```
What is the purpose of this if-statement? Won't it always be true?

genezys replied:
I did not study the whole call chain so I expected the parameter to be empty when not using the -c argument. Is it?

I think it should as that will allow docker-compose to fall back to its defaults and environment variables.

brikis98 replied:
Maybe I'm missing something, but aren't you declaring `yaml_file_param` on the line above this if-statement and setting it to an empty string? How could it have any other value?

genezys replied:
I am testing the value of `yaml_file_path` to build `yaml_file_param` by prepeding `-f` only if `yaml_file_path` is set.

If you have a prettier Bash syntax to do that, I am open!

brikis98 replied:
Oh jeez, my bad, I misread it as `yaml_file_param` in both the declaration and the if-statement. Not sure if that's just me or if `yaml_file_param` should be renamed to something like `docker_compose_param_yaml_file` :)

genezys replied:
Will do that, it will help differentiate both variables. Thanks!

brikis98 commented on src/docker-osx-dev:
```diff
@@ -992,28 +992,32 @@ function instructions {
-    done < "$yaml_file_path"
-  fi
+    else
+      if [[ "$line" = "volumes:" ]]; then
```
`==` is more readable.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -992,28 +992,32 @@ function instructions {
+  local line=""
+  while read line; do
+    if $in_volumes_block; then
+      if [[ "${line:0:2}" = "- " ]]; then
```
`==` is more readable

genezys replied:
That is not really a change I made. I just changed the existing indentation when removing an if. 

Shouldn't there be another PR for changing condition code styles?

brikis98 replied:
Oh, you're right. I must have missed that. OK, ignore that, and we'll do syntax clean up in a follow-up.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -992,28 +992,32 @@ function instructions {
+
+  local in_volumes_block=false
+  local line=""
+  while read line; do
```
It might help to add a comment here that shows an example of the `docker-compose config` output you're parsing.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -992,28 +992,32 @@ function instructions {
+    if $in_volumes_block; then
+      if [[ "${line:0:2}" = "- " ]]; then
+        local readonly path=$(echo $line | sed -ne "s/- \(\/[^:]*\):.*$/\1/p")
+        if [ ! -z "$path" ]; then
```
Please use `[[` and `]]`

brikis98 reviewed (changes requested):
Awesome, thank you! 

I left a few minor comments through out the code.

The only bigger thing would be to update the tests. All you'd need to do is to add a few example docker compose files under `test/resources`, one that uses `extends` and one that uses a named volume, and just ensure that the `load_paths_from_docker_compose` function parses those correctly. The test you update in this PR is a reasonable model to follow.

genezys commented:
Thanks for the feedback. Do you prefer I make another commit or amend the existing one?

brikis98 commented:
New commits would be great. Thanks!

brikis98 commented:
Oh, and let me know when you've updated things, as GitHub doesn't give me a notification for new commits being pushed.

genezys committed:
Add tests for extends & named volumes

genezys committed:
Rename variable for clarity + Fix if style

Prevent confusion between $yaml_file_path and $yaml_file_param

genezys committed:
Add example output when parsing docker-compose config

genezys commented:
@brikis98 I added both tests and an example output. I also fixed the style of the only new if that was added by my code, I did not touch the other ones. I also renamed the variable for clarity.

brikis98 reviewed (approved):
Fantastic, thank you!
